### PR TITLE
Add Curve-Amm example.

### PR DESCRIPTION
**Pull request by vivekvpandya**

### CODE REVIEW

File: fungible-token/src/lib.rs [hunk 1]
```diff
@@ -187,70 +199,13 @@ impl FungibleToken {
...
-    init:
-        input: InitConfig,
-    handle:
-        input: Action,
-        output: Event,
+        init:
+            input : InitConfig,
+        handle:
+            input : Action,
+            output : Event,
```
Comment (shamilsan) [inline fungible-token/src/lib.rs:208]: Prefer indents like in structs:
```suggestion
    init:
        input: InitConfig,
    handle:
        input: Action,
        output: Event,
```

File: fungible-token/src/lib.rs [hunk 2]
```diff
@@ -4,10 +4,11 @@
 #![no_std]
 #![feature(const_btree_new)]
 
-use codec::{Decode, Encode};
+// use codec::{Decode, Encode};
```
Comment (shamilsan) [inline fungible-token/src/lib.rs:7]: Dead code.
```suggestion
```

File: fungible-token/src/lib.rs [hunk 3]
```diff
@@ -4,10 +4,11 @@
...
 use gstd::{debug, exec, msg, prelude::*, ActorId};
 use primitive_types::H256;
-use scale_info::TypeInfo;
+// use sp_arithmetic::traits::One;
```
Comment (shamilsan) [inline fungible-token/src/lib.rs:10]: Dead code.
```suggestion
```

File: gear/gtest/spec/test_curve_amm.yaml [hunk 4]
```diff
@@ -0,0 +1,243 @@
...
+          token_accounts: "{1},{2},{3}"
+          amplification_coefficient : 1000
+          fee: 2
+
+fixtures:
+  - title: curve_amm
+    messages:
+      # add 10000 USDC to account 1 
+      - destination: 1
+        payload:
```
Comment (shamilsan) [inline gear/gtest/spec/test_curve_amm.yaml:40]: Repeating sections may be addressed using anchors.
```suggestion
        payload: &account-1
```

File: gear/gtest/spec/test_curve_amm.yaml [hunk 5]
```diff
@@ -0,0 +1,243 @@
...
+              account: 5FHneW46xGXgs5mUiveU4sbTyGBzmstUspZC92UhjJM694ty 
+              amount: 10000
+      # add 10000 USDT to account 1 
+      - destination: 2
+        payload:
+          kind: custom
+          value:
+            mint:
+              account: 5GrwvaEF5zXb26Fz9rcQpDWS57CtERHpNehXCPcNoHGKutQY
+              amount: 10000
```
Comment (shamilsan) [inline gear/gtest/spec/test_curve_amm.yaml:61]: See above about anchors.
```suggestion
        payload: *account-1
```

File: gear/gtest/spec/test_curve_amm.yaml [hunk 6]
```diff
@@ -0,0 +1,307 @@
...
+      #     value:
+      #       # balanceOf: 0x0400000000000000000000000000000000000000000000000000000000000000 
+      #       balanceOf: 5C9x3n5TQyVpXo5P7oojUXxQFQMRCQPomcZL5rndpBFReDFk 
+      # # check pool's balance for USDT
+      # - destination: 2
+      #   payload:
+      #     kind: custom
+      #     value:
+      #       balanceOf: 0x0400000000000000000000000000000000000000000000000000000000000000 
+      # add USDT, USDC liquidity from account 2 to pool
```
Comment (shamilsan) [inline gear/gtest/spec/test_curve_amm.yaml:112]: May we remove this code if it is not used?

```suggestion
```

File: gear/gtest/spec/test_fungible_token.yaml [hunk 7]
```diff
@@ -37,30 +37,40 @@ fixtures:
...
     expected:
       - log:
-          - destination: 0
+          - destination: 1000001
             payload:
               kind: custom
               value:
                 transfer:
-                  from: 0x0000000000000000000000000000000000000000000000000000000000000000
+                  from: 0x0000000000000000000000000000000000000000000000000000000000000000
```
Comment (shamilsan) [inline gear/gtest/spec/test_fungible_token.yaml:53]:
```suggestion
                  from: 0x0000000000000000000000000000000000000000000000000000000000000000
```

File: curve-amm/src/lib.rs [hunk 8]
```diff
@@ -0,0 +1,1166 @@
...
+
+#![no_std]
+#![feature(const_btree_new)]
+
+extern crate alloc;
+
+use alloc::collections::BTreeMap;
+use alloc::collections::BTreeSet;
+use alloc::vec;
+use alloc::vec::Vec;
```
Comment (shamilsan) [inline curve-amm/src/lib.rs:16]: `use` section can be combined to be shorter. Just for example.

```suggestion
use alloc::collections::{BTreeMap, BTreeSet};
use alloc::vec::{self, Vec};
```

But we usually include all module related stuff with the one `use` entry.

**Review by shamilsan — COMMENTED** [review]
My main concern is that the program is too big to be in one source file. I'd divide it into several modules.

**Comment by vivekvpandya** [comment]
> My main concern is that the program is too big to be in one source file. I'd divide it into several modules.

That kind of refactoring can be done in separate PR. Let's complete this.

**Comment by LouiseMedova** [comment]
@vivekvpandya  Maybe it's better to make Curve-Amm a library? Because when developing dex, curve liquidity pools will be used many times. I also think for a better understanding of program logic it would be better to make mathematical calculations in separate files

**Comment by vivekvpandya** [comment]
> @vivekvpandya Maybe it's better to make Curve-Amm a library? Because when developing dex, curve liquidity pools will be used many times. I also think for a better understanding of program logic it would be better to make mathematical calculations in separate files

Yes, we can put Action and Reply to a common lib, and perhaps few other functions, but still there will be these 3 core functions `add_liquilidity()` , `remove_liquidity()` and `exchange()` which requires to be implemented, no matter it is constant-product based AMM or StableSwap (Curve) AMM. 
Also this PR is very large , it also contains some refactoring in FungibleTokens so I request all to complete review and merge this. All big changes can be done incrementally.

**Comment by LouiseMedova** [comment]
I also think it's necessary to add reading of contracts state off-chain (in fungible-token and curve-amm). Then in fungible-token contract in functions `mint`, `burn`, `transfer`, `approve` you need to check the `msg::source()`. And maybe it's even better to reduce arguments in these functions (for example `sender` in `transfer` function can be removed and replaced with `msg::source()`). 
In curve amm there are comments `TODO` and usage of mutex, which is only needed to test the contract. But there should not be in the finished contract.

**Comment by shamilsan** [comment]
> Then in fungible-token contract in functions `mint`, `burn`, `transfer`, `approve` you need to check the `msg::source()`. And maybe it's even better to reduce arguments in these functions (for example `sender` in `transfer` function can be removed and replaced with `msg::source()`).

@vivekvpandya These things seem to be quite important to be fixed in this PR. Let's move the remaining to an issue and do it later.

File: curve-amm/src/lib.rs [hunk 9]
```diff
@@ -0,0 +1,1091 @@
...
+    ///  A * sum(x_i) * n^n + D = A * D * n^n + D^(n+1) / (n^n * prod(x_i))
+    /// ```
+    ///
+    /// Converging solution:
+    ///
+    /// ```pseudocode
+    /// D[j + 1] = (A * n^n * sum(x_i) - D[j]^(n+1) / (n^n * prod(x_i))) / (A * n^n - 1)
+    /// ```
+    #[allow(dead_code)]
+    pub fn get_d(&self, xp_f: &[FixedU128], ann_f: FixedU128) -> Option<FixedU128> {
```
[the remark this answers is not in the document]
Comment (vivekvpandya) [reply curve-amm/src/lib.rs:259]: these math functions are specific to stable-swap (i.e Curve.Finance) they can only be used in algorithm based on StableSwap.

File: curve-amm/src/lib.rs [hunk 10]
```diff
@@ -0,0 +1,1091 @@
...
+    InsufficientFunds,
+    /// Specified index is out of range
+    IndexOutOfRange,
+    /// The `AssetChecker` can use this error in case it can't provide better error
+    ExternalAssetCheckFailed,
+}
+
+/// Storage record type for a pool
+#[derive(Debug)]
+pub struct PoolInfo {
```
[the remark this answers is not in the document]
Comment (vivekvpandya) [reply curve-amm/src/lib.rs:200]: PoolInfo is small structure and for our use once pool is created things are mostly READ-ONLY. So I don't think it matters much to separate them.

**Comment by vivekvpandya** [comment]
Creating small PRs. Closing this.

**Pull request closed without merging**